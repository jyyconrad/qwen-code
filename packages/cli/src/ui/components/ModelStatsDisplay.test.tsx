/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { render } from 'ink-testing-library';
import { describe, it, expect, vi } from 'vitest';
import { ModelStatsDisplay } from './ModelStatsDisplay.js';
import * as SessionContext from '../contexts/SessionContext.js';
import { SessionMetrics } from '../contexts/SessionContext.js';

// 模拟上下文以提供受控数据用于测试
vi.mock('../contexts/SessionContext.js', async (importOriginal) => {
  const actual = await importOriginal<typeof SessionContext>();
  return {
    ...actual,
    useSessionStats: vi.fn(),
  };
});

const useSessionStatsMock = vi.mocked(SessionContext.useSessionStats);

const renderWithMockedStats = (metrics: SessionMetrics) => {
  useSessionStatsMock.mockReturnValue({
    stats: {
      sessionStartTime: new Date(),
      metrics,
      lastPromptTokenCount: 0,
      promptCount: 5,
    },

    getPromptCount: () => 5,
    startNewPrompt: vi.fn(),
  });

  return render(<ModelStatsDisplay />);
};

describe('<ModelStatsDisplay />', () => {
  it('当没有活动模型时应渲染 "no API calls" 消息', () => {
    const { lastFrame } = renderWithMockedStats({
      models: {},
      tools: {
        totalCalls: 0,
        totalSuccess: 0,
        totalFail: 0,
        totalDurationMs: 0,
        totalDecisions: { accept: 0, reject: 0, modify: 0 },
        byName: {},
      },
    });

    expect(lastFrame()).toContain(
      'No API calls have been made in this session.',
    );
    expect(lastFrame()).toMatchSnapshot();
  });

  it('如果没有模型具有相关数据，则不应显示条件行', () => {
    const { lastFrame } = renderWithMockedStats({
      models: {
        'gemini-2.5-pro': {
          api: { totalRequests: 1, totalErrors: 0, totalLatencyMs: 100 },
          tokens: {
            prompt: 10,
            candidates: 20,
            total: 30,
            cached: 0,
            thoughts: 0,
            tool: 0,
          },
        },
      },
      tools: {
        totalCalls: 0,
        totalSuccess: 0,
        totalFail: 0,
        totalDurationMs: 0,
        totalDecisions: { accept: 0, reject: 0, modify: 0 },
        byName: {},
      },
    });

    const output = lastFrame();
    expect(output).not.toContain('Cached');
    expect(output).not.toContain('Thoughts');
    expect(output).not.toContain('Tool');
    expect(output).toMatchSnapshot();
  });

  it('如果至少有一个模型具有数据，则应显示条件行', () => {
    const { lastFrame } = renderWithMockedStats({
      models: {
        'gemini-2.5-pro': {
          api: { totalRequests: 1, totalErrors: 0, totalLatencyMs: 100 },
          tokens: {
            prompt: 10,
            candidates: 20,
            total: 30,
            cached: 5,
            thoughts: 2,
            tool: 0,
          },
        },
        'gemini-2.5-flash': {
          api: { totalRequests: 1, totalErrors: 0, totalLatencyMs: 50 },
          tokens: {
            prompt: 5,
            candidates: 10,
            total: 15,
            cached: 0,
            thoughts: 0,
            tool: 3,
          },
        },
      },
      tools: {
        totalCalls: 0,
        totalSuccess: 0,
        totalFail: 0,
        totalDurationMs: 0,
        totalDecisions: { accept: 0, reject: 0, modify: 0 },
        byName: {},
      },
    });

    const output = lastFrame();
    expect(output).toContain('Cached');
    expect(output).toContain('Thoughts');
    expect(output).toContain('Tool');
    expect(output).toMatchSnapshot();
  });

  it('应正确显示多个模型的统计信息', () => {
    const { lastFrame } = renderWithMockedStats({
      models: {
        'gemini-2.5-pro': {
          api: { totalRequests: 10, totalErrors: 1, totalLatencyMs: 1000 },
          tokens: {
            prompt: 100,
            candidates: 200,
            total: 300,
            cached: 50,
            thoughts: 10,
            tool: 5,
          },
        },
        'gemini-2.5-flash': {
          api: { totalRequests: 20, totalErrors: 2, totalLatencyMs: 500 },
          tokens: {
            prompt: 200,
            candidates: 400,
            total: 600,
            cached: 100,
            thoughts: 20,
            tool: 10,
          },
        },
      },
      tools: {
        totalCalls: 0,
        totalSuccess: 0,
        totalFail: 0,
        totalDurationMs: 0,
        totalDecisions: { accept: 0, reject: 0, modify: 0 },
        byName: {},
      },
    });

    const output = lastFrame();
    expect(output).toContain('gemini-2.5-pro');
    expect(output).toContain('gemini-2.5-flash');
    expect(output).toMatchSnapshot();
  });

  it('应处理大值而不换行或重叠', () => {
    const { lastFrame } = renderWithMockedStats({
      models: {
        'gemini-2.5-pro': {
          api: {
            totalRequests: 999999999,
            totalErrors: 123456789,
            totalLatencyMs: 9876,
          },
          tokens: {
            prompt: 987654321,
            candidates: 123456789,
            total: 999999999,
            cached: 123456789,
            thoughts: 111111111,
            tool: 222222222,
          },
        },
      },
      tools: {
        totalCalls: 0,
        totalSuccess: 0,
        totalFail: 0,
        totalDurationMs: 0,
        totalDecisions: { accept: 0, reject: 0, modify: 0 },
        byName: {},
      },
    });

    expect(lastFrame()).toMatchSnapshot();
  });

  it('应正确显示单个模型', () => {
    const { lastFrame } = renderWithMockedStats({
      models: {
        'gemini-2.5-pro': {
          api: { totalRequests: 1, totalErrors: 0, totalLatencyMs: 100 },
          tokens: {
            prompt: 10,
            candidates: 20,
            total: 30,
            cached: 5,
            thoughts: 2,
            tool: 1,
          },
        },
      },
      tools: {
        totalCalls: 0,
        totalSuccess: 0,
        totalFail: 0,
        totalDurationMs: 0,
        totalDecisions: { accept: 0, reject: 0, modify: 0 },
        byName: {},
      },
    });

    const output = lastFrame();
    expect(output).toContain('gemini-2.5-pro');
    expect(output).not.toContain('gemini-2.5-flash');
    expect(output).toMatchSnapshot();
  });
});