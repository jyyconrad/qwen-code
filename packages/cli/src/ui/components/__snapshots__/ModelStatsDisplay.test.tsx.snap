// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ModelStatsDisplay /> > should display a single model correctly 1`] = `
"╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                                  │
│  Model Stats For Nerds                                                                           │
│                                                                                                  │
│  Metric                      gemini-2.5-pro                                                      │
│  ──────────────────────────────────────────────────────────────────────────────────────────────  │
│  API                                                                                             │
│  Requests                    1                                                                   │
│  Errors                      0 (0.0%)                                                            │
│  Avg Latency                 100ms                                                               │
│                                                                                                  │
│  Tokens                                                                                          │
│  Total                       30                                                                  │
│    ↳ Prompt                  10                                                                  │
│    ↳ Cached                  5 (50.0%)                                                           │
│    ↳ Thoughts                2                                                                   │
│    ↳ Tool                    1                                                                   │
│    ↳ Output                  20                                                                  │
│                                                                                                  │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯"
`;

exports[`<ModelStatsDisplay /> > should display conditional rows if at least one model has data 1`] = `
"╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                                  │
│  Model Stats For Nerds                                                                           │
│                                                                                                  │
│  Metric                      gemini-2.5-pro        gemini-2.5-flash                              │
│  ──────────────────────────────────────────────────────────────────────────────────────────────  │
│  API                                                                                             │
│  Requests                    1                     1                                             │
│  Errors                      0 (0.0%)              0 (0.0%)                                      │
│  Avg Latency                 100ms                 50ms                                          │
│                                                                                                  │
│  Tokens                                                                                          │
│  Total                       30                    15                                            │
│    ↳ Prompt                  10                    5                                             │
│    ↳ Cached                  5 (50.0%)             0 (0.0%)                                      │
│    ↳ Thoughts                2                     0                                             │
│    ↳ Tool                    0                     3                                             │
│    ↳ Output                  20                    10                                            │
│                                                                                                  │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯"
`;

exports[`<ModelStatsDisplay /> > should display stats for multiple models correctly 1`] = `
"╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                                  │
│  Model Stats For Nerds                                                                           │
│                                                                                                  │
│  Metric                      gemini-2.5-pro        gemini-2.5-flash                              │
│  ──────────────────────────────────────────────────────────────────────────────────────────────  │
│  API                                                                                             │
│  Requests                    10                    20                                            │
│  Errors                      1 (10.0%)             2 (10.0%)                                     │
│  Avg Latency                 100ms                 25ms                                          │
│                                                                                                  │
│  Tokens                                                                                          │
│  Total                       300                   600                                           │
│    ↳ Prompt                  100                   200                                           │
│    ↳ Cached                  50 (50.0%)            100 (50.0%)                                   │
│    ↳ Thoughts                10                    20                                            │
│    ↳ Tool                    5                     10                                            │
│    ↳ Output                  200                   400                                           │
│                                                                                                  │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯"
`;

exports[`<ModelStatsDisplay /> > should handle large values without wrapping or overlapping 1`] = `
"╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                                  │
│  Model Stats For Nerds                                                                           │
│                                                                                                  │
│  Metric                      gemini-2.5-pro                                                      │
│  ──────────────────────────────────────────────────────────────────────────────────────────────  │
│  API                                                                                             │
│  Requests                    999,999,999                                                         │
│  Errors                      123,456,789 (12.3%)                                                 │
│  Avg Latency                 0ms                                                                 │
│                                                                                                  │
│  Tokens                                                                                          │
│  Total                       999,999,999                                                         │
│    ↳ Prompt                  987,654,321                                                         │
│    ↳ Cached                  123,456,789 (12.5%)                                                 │
│    ↳ Thoughts                111,111,111                                                         │
│    ↳ Tool                    222,222,222                                                         │
│    ↳ Output                  123,456,789                                                         │
│                                                                                                  │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯"
`;

exports[`<ModelStatsDisplay /> > should not display conditional rows if no model has data for them 1`] = `
"╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                                  │
│  Model Stats For Nerds                                                                           │
│                                                                                                  │
│  Metric                      gemini-2.5-pro                                                      │
│  ──────────────────────────────────────────────────────────────────────────────────────────────  │
│  API                                                                                             │
│  Requests                    1                                                                   │
│  Errors                      0 (0.0%)                                                            │
│  Avg Latency                 100ms                                                               │
│                                                                                                  │
│  Tokens                                                                                          │
│  Total                       30                                                                  │
│    ↳ Prompt                  10                                                                  │
│    ↳ Output                  20                                                                  │
│                                                                                                  │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯"
`;

exports[`<ModelStatsDisplay /> > should render "no API calls" message when there are no active models 1`] = `
"╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                                  │
│  No API calls have been made in this session.                                                    │
│                                                                                                  │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯"
`;

exports[`<ModelStatsDisplay /> > 如果没有模型具有相关数据，则不应显示条件行 1`] = `
"╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                                  │
│  模型统计信息（技术版）                                                                          │
│                                                                                                  │
│  指标                        gemini-2.5-pro                                                      │
│  ──────────────────────────────────────────────────────────────────────────────────────────────  │
│  API                                                                                             │
│  请求次数                    1                                                                   │
│  错误数                      0 (0.0%)                                                            │
│  平均延迟                    100ms                                                               │
│                                                                                                  │
│  Token                                                                                           │
│  总计                        30                                                                  │
│    ↳ 提示词                  10                                                                  │
│    ↳ 输出                    20                                                                  │
│                                                                                                  │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯"
`;

exports[`<ModelStatsDisplay /> > 应处理大值而不换行或重叠 1`] = `
"╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                                  │
│  模型统计信息（技术版）                                                                          │
│                                                                                                  │
│  指标                        gemini-2.5-pro                                                      │
│  ──────────────────────────────────────────────────────────────────────────────────────────────  │
│  API                                                                                             │
│  请求次数                    999,999,999                                                         │
│  错误数                      123,456,789 (12.3%)                                                 │
│  平均延迟                    0ms                                                                 │
│                                                                                                  │
│  Token                                                                                           │
│  总计                        999,999,999                                                         │
│    ↳ 提示词                  987,654,321                                                         │
│    ↳ 缓存                    123,456,789 (12.5%)                                                 │
│    ↳ 思考内容                111,111,111                                                         │
│    ↳ 工具                    222,222,222                                                         │
│    ↳ 输出                    123,456,789                                                         │
│                                                                                                  │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯"
`;

exports[`<ModelStatsDisplay /> > 应正确显示单个模型 1`] = `
"╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                                  │
│  模型统计信息（技术版）                                                                          │
│                                                                                                  │
│  指标                        gemini-2.5-pro                                                      │
│  ──────────────────────────────────────────────────────────────────────────────────────────────  │
│  API                                                                                             │
│  请求次数                    1                                                                   │
│  错误数                      0 (0.0%)                                                            │
│  平均延迟                    100ms                                                               │
│                                                                                                  │
│  Token                                                                                           │
│  总计                        30                                                                  │
│    ↳ 提示词                  10                                                                  │
│    ↳ 缓存                    5 (50.0%)                                                           │
│    ↳ 思考内容                2                                                                   │
│    ↳ 工具                    1                                                                   │
│    ↳ 输出                    20                                                                  │
│                                                                                                  │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯"
`;

exports[`<ModelStatsDisplay /> > 应正确显示多个模型的统计信息 1`] = `
"╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                                  │
│  模型统计信息（技术版）                                                                          │
│                                                                                                  │
│  指标                        gemini-2.5-pro        gemini-2.5-flash                              │
│  ──────────────────────────────────────────────────────────────────────────────────────────────  │
│  API                                                                                             │
│  请求次数                    10                    20                                            │
│  错误数                      1 (10.0%)             2 (10.0%)                                     │
│  平均延迟                    100ms                 25ms                                          │
│                                                                                                  │
│  Token                                                                                           │
│  总计                        300                   600                                           │
│    ↳ 提示词                  100                   200                                           │
│    ↳ 缓存                    50 (50.0%)            100 (50.0%)                                   │
│    ↳ 思考内容                10                    20                                            │
│    ↳ 工具                    5                     10                                            │
│    ↳ 输出                    200                   400                                           │
│                                                                                                  │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯"
`;
